src/registrar.js

```diff
@@ -0,0 +1,165 @@
+'use strict'
+
+const assert = require('assert')
+const debug = require('debug')
+const log = debug('libp2p:peer-store')
+log.error = debug('libp2p:peer-store:error')
+const errCode = require('err-code')
+
+const { Connection } = require('libp2p-interfaces/src/connection')
+const PeerInfo = require('peer-info')
+const MulticodecTopology = require('./connection-manager/topology')
+
+/**
+ * Responsible for notifying registered protocols of events in the network.
+ */
+class Registrar {
+  /**
+   * @param {PeerStore} peerStore
+   * @constructor
+   */
+  constructor (peerStore) {
+    this.peerStore = peerStore
+
+    /**
+     * Map of connections per peer
+     * TODO: this should be handled by connectionManager
+     * @type {Map<string, conn>}
+     */
+    this.connections = new Map()
+
+    /**
+     * Map of topologies per multicodec
+     *
+     * @type {Map<string, object>}
+     */
+    this.multicodecTopologies = new Map()
+
+    this._handle = undefined
+  }
+
+  get handle () {
+    return this._handle
+  }
+
+  set handle (handle) {
+    this._handle = handle
+  }
+
+  /**
+   * Add a new connected peer to the record
+   * TODO: this should live in the ConnectionManager
+   * @param {PeerInfo} peerInfo
+   * @param {Connection} conn
+   * @returns {void}
+   */
+  onConnect (peerInfo, conn) {
+    assert(PeerInfo.isPeerInfo(peerInfo), 'peerInfo must be an instance of peer-info')
+    assert(Connection.isConnection(conn), 'conn must be an instance of interface-connection')
+
+    this.connections.set(peerInfo.id.toB58String(), conn)
```

Comment:
This could be a problem, there may be more than 1 connection to a peer. It's not the desired behavior but it can happen and we don't want to lose tracking of this. The connection Map should probably take an Array of connections for now.